client-http/src/main/java/com/cloudera/livy/client/http/HttpClient.java

```diff
@@ -0,0 +1,156 @@
+  private final int sessionId;
+  private final ScheduledExecutorService executor;
+  private final Serializer serializer;
+
+  private boolean stopped;
+
+  HttpClient(URI uri, HttpConf httpConf) {
+    this.config = httpConf;
+    this.conn = new LivyConnection(uri, httpConf);
+    this.stopped = false;
+
+    try {
+      long timeout = config.getTimeAsMs(SESSION_CREATE_TIMEOUT);
+      Map<String, String> sessionConf = new HashMap<>();
+      for (Map.Entry<String, String> e : config) {
+        sessionConf.put(e.getKey(), e.getValue());
+      }
+
+      ClientMessage create = new CreateClientRequest(timeout, sessionConf);
+      this.sessionId = conn.post(create, SessionInfo.class, "/").id;
+    } catch (Exception e) {
+      throw propagate(e);
+    }
+
+    // Because we only have one connection to the server, we don't need more than a single
+    // threaded executor here.
+    this.executor = Executors.newSingleThreadScheduledExecutor(new ThreadFactory() {
+      @Override
+      public Thread newThread(Runnable r) {
+        Thread t = new Thread(r, "HttpClient-" + sessionId);
+        t.setDaemon(true);
+        return t;
+      }
+    });
+
+    this.serializer = new Serializer();
+  }
+
+  @Override
+  public <T> JobHandle<T> submit(Job<T> job) {
+    return sendJob("submit-job", job);
+  }
+
+  @Override
+  public <T> Future<T> run(Job<T> job) {
+    return sendJob("run-job", job);
+  }
+
+  @Override
+  public synchronized void stop() {
+    if (!stopped) {
+      executor.shutdownNow();
+      try {
+        conn.delete(Map.class, "/%s", sessionId);
+      } catch (Exception e) {
+        throw propagate(e);
+      } finally {
+        try {
+          conn.close();
+        } catch (Exception e) {
+          // Ignore.
+        }
+      }
+      stopped = true;
+    }
+  }
+
+  @Override
+  public Future<?> addJar(final URI uri) {
+    return addResource("add-jar", uri);
+  }
+
+  @Override
+  public Future<?> addFile(URI uri) {
+    return addResource("add-file", uri);
+  }
+
+  private Future<?> addResource(final String command, final URI resource) {
+    if (resource.getScheme() == null || resource.getScheme() == "file") {
+      throw new IllegalArgumentException("Local resources are not yet supported: " + resource);
+    }
+
+    Callable<Void> task = new Callable<Void>() {
+      @Override
+      public Void call() throws Exception {
+        ClientMessage msg = new AddResource(resource.toString());
+        conn.post(msg, Void.class, "/%d/%s", sessionId, command);
+        return null;
+      }
+    };
+    return executor.submit(task);
+  }
+
+  private <T> JobHandleImpl<T> sendJob(final String command, Job<T> job) {
+    final ByteBuffer serializedJob = serializer.serialize(job);
+    JobHandleImpl<T> handle = new JobHandleImpl<T>(config, conn, executor, serializer);
+    handle.start(sessionId, command, serializedJob);
+    return handle;
+  }
+
+  private RuntimeException propagate(Exception cause) {
```

Comment:
There are a couple of other places in the existing codebase where I felt it might be useful to add a `LivyException` - that might be beneficial here too.